web/packages/teleport/src/components/hooks/useKeyBasedPagination.ts

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useState, useEffect, useRef } from 'react';
+import useAttempt, { Attempt } from 'shared/hooks/useAttemptNext';
+
+import {
+  ResourcesResponse,
+  ResourceFilter,
+  resourceFilterToHookDeps,
+} from 'teleport/services/agents';
+import { UrlResourcesParams } from 'teleport/config';
+
+/**
+ * Supports fetching more data from the server when more data is available. Pass
+ * a `fetchFunc` that retrieves a single batch of data. After the initial
+ * request, the server is expected to return a `startKey` field that denotes the
+ * next `startKey` to use for the next request.
+ *
+ * This hook is intended to be used in tandem with the `useInfiniteScroll` hook.
+ * See its documentation for an example.
+ */
+export function useKeyBasedPagination<T>({
+  fetchFunc,
+  clusterId,
+  filter,
+  initialFetchSize = 30,
+  fetchMoreSize = 20,
+}: Props<T>): State<T> {
+  const abortController = useRef<AbortController | null>(null);
+  const { attempt, setAttempt } = useAttempt();
+  const [finished, setFinished] = useState(false);
+  const [resources, setResources] = useState<T[]>([]);
+  const [startKey, setStartKey] = useState(null);
+
+  useEffect(() => {
+    abortController.current?.abort();
+    abortController.current = null;
+
+    setAttempt({ status: '', statusText: '' });
+    setFinished(false);
+    setResources([]);
+    setStartKey(null);
+  }, [clusterId, ...resourceFilterToHookDeps(filter)]);
+
+  // This is an additional countermeasure to prevent fetching if `fetch` is
+  // called multiple times per React state reconciliation cycle.
+  let fetchCalled = false;
```

Comment:
This can't be relied upon, since it will be re-evaluated each time we re-render. If you need to keep state between re-renders without re-rendering each time it's updated, you typically want a [`useRef`](https://react.dev/reference/react/useRef).

Reply:
That's reflected in the `attempt.status` field, so in case of re-renders, we're covered. Yes, I could `useRef` here, but I'm exactly interested in cases where this function may be called multiple times per render cycle. This is because — unless you prove me wrong, which I'd be glad about, since it simplifies things — React doesn't give me any guarantees that my intersection observer won't trigger more than once before React recomputes state. In such case, it could easily call `fetch` two times.

I'll update the comment, since perhaps it's not clear enough.

Reply:
> I'm exactly interested in cases where this function may be called multiple times per render cycle. This is because — unless you prove me wrong, which I'd be glad about, since it simplifies things — React doesn't give me any guarantees that my intersection observer won't trigger more than once before React recomputes state. In such case, it could easily call `fetch` two times.

Is this to avoid fetching if you're already fetching? This might be helpful. https://react.dev/learn/you-might-not-need-an-effect#fetching-data

`useAsync` also includes a mechanism to automatically discard old data.

I agree with Isaiah that this most probably should be replaced by some other mechanism.

I remember these resources helped me wrap my head around it [the last time I had to deal with stuff like this](https://github.com/gravitational/teleport/pull/26143):

* https://react.dev/learn/queueing-a-series-of-state-updates
* https://javascript.plainenglish.io/batch-updates-in-react-17-or-earlier-versions-5f76d58e3af1

Reply:
> I'll update the comment, since perhaps it's not clear enough.

My mistake, now that I'm looking at it again, the wording of the comment is clear. That said, it still deserves a deeper explanation for why such a thing needs to be accounted for, given that it's an unusual constraint. I also find this to be another point in favor of marrying this and `useInfiniteScroll`, as it's awkward to be accounting for the properties of a mechanism that's "nowhere in sight".

Reply:
I rewrote this area completely, it's a bit less hacky right now. (I think everything I used is actually in line with React guidelines.)

@ravicious `useAsync` is a great alternative, but since I already got this done, I'd like to keep it as is. Next time I see a similar pain, I'll consider using the `react-async` library.